### PR TITLE
[#77] Add support for glob patterns to `ignored` and `notScanned`

### LINKED ISSUE #77: Add support for glob patterns to `ignored` and `notScanned`
# Clarification and motivation

The `virtualFiles` config option supports glob patterns.

On the other hand, `ignored` only supports exact matches and `notScanned` matches on prefixes.

We should change these two config options to also support glob patterns like `virtualFiles`.

# Acceptance criteria

All 3 config options work similarly.

### PR DESCRIPTION
**Pull request by Sereja313**
## Description

Problem: The `virtualFiles` config option supports glob patterns. On the
other hand, `ignored` only supports exact matches and `notScanned`
mathches on prefixes. There is also a bug where `ignored` does not
ignore files if they contain broken xrefcheck annotations.

Solution: Add support for glob patterns to `ignored` and
`notScanned`. Filter ignored files before parsing their contents.

## Related issue(s)

Fixes #77

## :white_check_mark: Checklist for your Pull Request

Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](/README.md)
    - Haddock

- Public contracts
  - [ ] Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - [x] I added an entry to the [changelog](/CHANGES.md) if my changes are visible to the users
        and
  - [ ] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](/docs/code-style.md).

### CODE REVIEW

File: tests/golden/check-ignored/check-ignored.bats [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  run xrefcheck --ignored 'to-ignore/*'
+
+  assert_output --partial "All repository links are valid."
+}
+
+@test "Ignore file with broken xrefcheck annotation: config file" {
+  run xrefcheck --config ./config-ignored.yaml
+
+  assert_output --partial "All repository links are valid."
+}
```
Comment (dcastro) [inline tests/golden/check-ignored/check-ignored.bats:28]: Great tests! :+1: can you add some for `notScanned` as well please?

File: tests/golden/check-ignored/check-ignored.bats [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+load '../helpers/bats-support/load'
+load '../helpers/bats-assert/load'
+load '../helpers'
+
+
+@test "Ignore file with broken xrefcheck annotation: full path" {
+  run xrefcheck --ignored ./to-ignore/broken_annotation.md
+
+  assert_output --partial "All repository links are valid."
+}
```
Comment (dcastro) [inline tests/golden/check-ignored/check-ignored.bats:16]: For completeness, can you add a test for "directory path" with just `to-ignore` please?

**Review by dcastro — CHANGES_REQUESTED** [review]

File: tests/golden/check-ignored/check-ignored.bats (hunk 1 above)
Comment (Sereja313) [reply tests/golden/check-ignored/check-ignored.bats:34]: Oh, while adding these tests, I noticed that `notScanned` doesn't actually filter directories... Thanks for helping me notice that!  :smile:

**Review by dcastro — CHANGES_REQUESTED** [review]

File: src/Xrefcheck/Verify.hs [hunk 3]
```diff
@@ -230,7 +230,8 @@ verifyRepo
     = do
   let toScan = do
         (file, fileInfo) <- M.toList repoInfo
-        guard . not $ matchesGlobPatterns root vcNotScanned file
+        let isNotScanned = matchesGlobPatterns root vcNotScanned
+        guard . not $ isNotScanned file || isNotScanned (takeDirectory file)
```
[the remark this answers is not in the document]
Comment (Sereja313) [reply src/Xrefcheck/Verify.hs:234]: Oh, indeed, this wouldn't work on nested directories.
We are traversing the directory tree in `filterExcludedDirs`, that's why `ignored` works on subdirectories
https://github.com/serokell/xrefcheck/blob/4fd79cb12ec959cb2dfbfd3f6ce3762da88b43e0/src/Xrefcheck/Scan.hs#L91-L101

**Review by dcastro — APPROVED** [review]
Great job! :D

**Pull request merged**